### ISSUE TITLE
Query callbacks should fire asynchronously

### ISSUE THREAD

**Issue by rkirov** [angular/angular#4186]
The following usecase currently throws an error, because the directives are modified after they are dirtychecked.

```javascript
@Component(/* binds foo to template */)
class Foo {...}

@Component(...)
consturctor(public @Query(Foo) query) {
   this.query.addCallback(this.query.forEach(q) => q.foo = false);
}
```
However, if the callbacks are fired asynchronously within the angular zone, this will be valid code. There is a possibility for infinite async cycle in that case, but I don't think that is a big concern. Such possibility exists even with regular Promises.

Note, this might naturally fall out from moving Queries to Rx (TS) / Stream (dart).
cc @vsavkin

**Comment by mhevery** [comment]
@rkirov This is not valid code??? Could you rewrite? `addCallback` takes a function.

**Comment by pkozlowski-opensource** [comment]
@mhevery @rkirov we've got a use case, actually: let's say I've got a component that updates its internal state based on what happens to its children.

More specifically: say I've got a `<carousel>` component with several `<slide>` children. I want to observe a list of slides to update carousel's state (for example to make sure that the current slide is not the deleted one, update navigation indicators, display a no of slides etc.). In this case changes to the carousel model done in response to slides' collection change and not reflected in the UI.

I don't know, maybe there is a better way of solving use-cases like this one, but @mlaval and I weren't sure what could be an alternative approach.
